### DISCUSSION TITLE
Google calendar entity model other / upcoming events

### DISCUSSION THREAD

**Discussion by jampez77** [home-assistant/architecture#567] [Entity Models] [answered]
I've created to PRs ([47049](https://github.com/home-assistant/core/pull/47049/) / [50856](https://github.com/home-assistant/core/pull/50856)) that require slight changes to the Google calendar entity model.

This is the current model:

```
{            
    "message": event["message"],            
    "all_day": event["all_day"],            
    "start_time": event["start"],            
    "end_time": event["end"],            
    "location": event["location"],            
    "description": event["description"],        
}
```

My proposal is that we append `other_events` and / or `upcoming_events` to that model when there are more events to add.

Definitions:

Other Events: Any event(s) happening within the start and end time of the current event.
Upcoming Events: All events happening within the next 7 days.

**Comment by MartinHjelmare** [comment]
Note that according to the PR you want to change not only the Google calendar but the base calendar entity.

**Reply by jampez77** [reply]
Yes, I'm sure this is just an indication that I don't understand what I need to change but fundamentally the changes I make to the Google calendar integration don't show up in the frontend unless I also update the base calendar entity.

**Reply by MartinHjelmare** [reply]
The frontend calendar feature is made compatible with the base calendar entity, so yes that is expected.

**Comment by allenporter** [comment]
Naive question: That entry there looks like it represents and event.  Would upcoming and other events also just be represented as additional separate event entries returned? It would help to understand more detail behind this as it feels like this proposal is a workaround for a different limitation, but I can't tell.

**Reply by jampez77** [reply]
Sorry, I'm not sure what you mean.

Do you mean the other / upcoming events also showing the new attributes?

**Reply by allenporter** [reply]
I'm asking for more detail because I don't understand.

I read that as the model itself represents an event.  So if you have event #1 and associated/other events A and B, the proposal here is to return as properties of #1 if I understand.

Ad an alternative: does it work to return A and B as new events #2 and #3?

**Reply by jampez77** [reply]
Ah ok, so it would be the first scenario. So for example:

```
{
   "name": "EVENT 1",
   "start": "TODAY 09:00",
   "end":"TODAY 19:00",
   "other_events": [
           {
                "name": "EVENT A",
                "start": "TODAY 10:00",
                "end":"TODAY 12:00"
           }
   ],
  "upcoming_events": [
           {
                "name": "EVENT B",
                "start": "TOMORROW 10:00",
                "end":"TOMORROW 22:00"
           }
  ]
}
```

**Reply by allenporter** [reply]
Sorry I wasn't clear: I understand what the proposal is, but not sure *why* or what the goal is.

 Compared to an alternative using the existing data model which I would assume already returns the events as a flat list:
```
{
   "name": "EVENT 1",
   "start": "TODAY 09:00",
   "end":"TODAY 19:00",
},
{
   "name": "EVENT A",
   "start": "TODAY 10:00",
   "end":"TODAY 12:00",
},
{
   "name": "EVENT B",
   "start": "TOMORROW 10:00",
   "end":"TOMORROW 22:00",
}
```
That is, the data model can represent all of this information already.

It would help to understand more detail, what is the motivation, what is the ultimate goal, what are we trying to accomplish, etc.

**Reply by jampez77** [reply]
Ah ok, I'm with you now!

Basically the goal is to expose more than just the current event from a Google calendar.

The backstory is that I want to build an automation that will add a birthday card item to a shopping list when a calendar event (someones birthday) is so many days in the future.

If there is a better way of doing things i'm open to it. I guess I updated the model as having nested other / upcoming events as it made sense to me. No special reason at all.

**Reply by allenporter** [reply]
I'm not sure of an approach that will work for that yet, but i think sticking to the flat list would be the way to go. (e.g. ensuring that events within 7 days are returned as part of the events).

The part that is a bit difficult to work with is that a calendar event is only `on` when it is active. In this case you want it to be `on` before the event happens.   It looks like the proposed way to do this is here: https://www.home-assistant.io/integrations/calendar.google/#calendar-configuration where you add an offset to trigger the event early.

**Reply by jampez77** [reply]
I'm sorry I'm not sure I understand. I don't want the calendar entity to be `on` before the event happens. I haven't altered that so it should be `on` or `off` depending on the current event and nothing else.

I just want to have access to events before they happen as well as being shown other events happening during the same time.

A nested list strikes me as a more intuitive model.

**Reply by allenporter** [reply]
Oh i assumed the binary sensor being `on` is the primary way to trigger a calendar automation. How do automations work in this context?

If I understand this right a single event also returns all upcoming events so there's kind of a blow up here of every event having a copy of every other event embedded inside?  So really the model now looks like this:
```
{
   "name": "EVENT 1",
   "start": "TODAY 09:00",
   "end":"TODAY 19:00",
   "other_events": [
           {
                "name": "EVENT A",
                "start": "TODAY 10:00",
                "end":"TODAY 12:00"
           }
   ],
  "upcoming_events": [
           {
                "name": "EVENT B",
                "start": "TOMORROW 10:00",
                "end":"TOMORROW 22:00"
           }
  ]
},
{
   "name": "EVENT A",
   "start": "TODAY 10:00",
   "end":"TODAY 12:00",
   "other_events": [
           {
                "name": "EVENT 1",
                "start": "TODAY 09:00",
                "end":"TODAY 19:00"
           }
   ],
  "upcoming_events": [
           {
                "name": "EVENT B",
                "start": "TOMORROW 10:00",
                "end":"TOMORROW 22:00"
           }
  ]
},
{
   "name": "EVENT B",
   "start": "TOMORROW 10:00",
   "end":"TOMORROW 12:00",
}
```

**Reply by jampez77** [reply]
I think it is probably the primary way and the way that works won't be impacted by these changes.

The calendar iteration only event shows a single object (the next event). It's not a list like you think. 

so the model will only be the single objects with other / upcoming events appended where they exist.

Like this:
```
{
   "name": "EVENT 1",
   "start": "TODAY 09:00",
   "end":"TODAY 19:00",
   "other_events": [
           {
                "name": "EVENT A",
                "start": "TODAY 10:00",
                "end":"TODAY 12:00"
           }
   ],
  "upcoming_events": [
           {
                "name": "EVENT B",
                "start": "TOMORROW 10:00",
                "end":"TOMORROW 22:00"
           }
  ]
}
```

**Reply by MartinHjelmare** [reply]
~~Why do you want to access information about the other/upcoming events? What will you do with that information?~~

I found your answer above:
> The backstory is that I want to build an automation that will add a birthday card item to a shopping list when a calendar event (someones birthday) is so many days in the future.

**Reply by MartinHjelmare** [reply, marked as answer]
My suggestion is that we add a new trigger platform for the calendar integration which will allow users to define automation triggers searching for matching events within a time frame. When there's a match the trigger will fire and the automation will run.

**Reply by MartinHjelmare** [reply]
This is now available.

**Reply by jampez77** [reply]
Hey, thanks for the update. How exactly would I do this? I can see the trigger in Automations but there is nothing I can see about specifying a time frame or matching events.

**Reply by allenporter** [reply]
Hi, check out https://www.home-assistant.io/integrations/calendar/#automation and see the example 'Calendar Event Light Schedule' which looks at matching specific types of events.

**Reply by MartinHjelmare** [reply]
Use a condition and check the calendar trigger template variables as needed:

https://www.home-assistant.io/docs/automation/templating/#calendar

**Reply by jampez77** [reply]
Hey @allenporter thanks for pointing me in that direction!. However this doesn't appear to do what is intended, The automation runs regardless of what text I put in the template condition, unless i'm missing something I think it could be broken.

**Reply by MartinHjelmare** [reply]
This isn't the place for troubleshooting automations. I suggest moving to discord or the community forum.
